File: mobile/lib/utils/distance.dart
import 'dart:math';

/// Haversine distance — straight-line km between two coordinates.
double? distanceKm(double? aLat, double? aLng, double? bLat, double? bLng) {
  if (aLat == null || aLng == null || bLat == null || bLng == null) return null;
  const r = 6371.0;
  double toRad(double d) => d * pi / 180;
  final dLat = toRad(bLat - aLat);
  final dLng = toRad(bLng - aLng);
  final sa = sin(dLat / 2);
  final sb = sin(dLng / 2);
  final x = sa * sa + cos(toRad(aLat)) * cos(toRad(bLat)) * sb * sb;
  return ((r * 2 * asin(min(1, sqrt(x)))) * 10).round() / 10;
}

String formatDistance(double? km) {
  if (km == null) return '—';
  if (km < 1) return '${(km * 1000).round()} m away';
  return '$km km away';
}

File: web/src/lib/distance.ts
// Haversine distance — straight-line km between two coordinates.
export function distanceKm(
  a: { lat: number; lng: number } | undefined | null,
  b: { lat: number; lng: number } | undefined | null
): number | null {
  if (!a || !b || a.lat == null || b.lat == null) return null;
  const R = 6371;
  const toRad = (d: number) => (d * Math.PI) / 180;
  const dLat = toRad(b.lat - a.lat);
  const dLng = toRad(b.lng - a.lng);
  const sa = Math.sin(dLat / 2);
  const sb = Math.sin(dLng / 2);
  const x = sa * sa + Math.cos(toRad(a.lat)) * Math.cos(toRad(b.lat)) * sb * sb;
  return Math.round(R * 2 * Math.asin(Math.min(1, Math.sqrt(x))) * 10) / 10;
}

export function formatDistance(km: number | null): string {
  if (km == null) return "—";
  if (km < 1) return `${Math.round(km * 1000)} m away`;
  return `${km} km away`;
}
